### PR TITLE
[SPARK-29801][ML] ML models unify toString method

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1,ML models should extend toString method to expose basic information.
Current some algs (GBT/RF/LoR) had done this, while others not yet.
2,add `val numFeatures` in `BisectingKMeansModel`/`GaussianMixtureModel`/`KMeansModel`/`AFTSurvivalRegressionModel`/`IsotonicRegressionModel`

### Why are the changes needed?
ML models should extend toString method to expose basic information.

### Does this PR introduce any user-facing change?
yes

### How was this patch tested?
existing testsuites

### CODE REVIEW

**Comment by zhengruifeng** [comment]
For example, KMeansModel.toString should expose some basice information like LoR, but it only contains Estimator's uid now.

```scala
scala> val kmeans = new KMeans().setK(2).setSeed(1L)
kmeans: org.apache.spark.ml.clustering.KMeans = kmeans_4d760be410ff

scala> val kmeans_model = kmeans.fit(dataset)
kmeans_model: org.apache.spark.ml.clustering.KMeansModel = kmeans_4d760be410ff

scala> val lr = new LogisticRegression().setMaxIter(1)
lr: org.apache.spark.ml.classification.LogisticRegression = logreg_6523109640c2

scala> val lrm = lr.fit(dataset)
lrm: org.apache.spark.ml.classification.LogisticRegressionModel = LogisticRegressionModel: uid = logreg_6523109640c2, numClasses = 6, numFeatures = 3
```

**Review by srowen — COMMENTED** [review]
That seems fine. Are there any existing toString methods that need to be made consistent with this?

**Comment by huaxingao** [comment]
I recall the problem we had earlier on Python side when setting params for ML models, for example: 
```
    >>> gm = GaussianMixture(k=3, tol=0.0001, seed=10)
    >>> model = gm.fit(df)
    >>> model.setPredictionCol("newPrediction")
    GaussianMixture...
```
I had to put ```GaussianMixture...``` in the last line because ```GaussianMixtureModel...``` would fail.
Now since the ```toString``` is available in ML models on scala side, I will open a jira to fix the python side problem by adding a ```__repr__(self)``` method in the ML models
```
    def __repr__(self):
        return self._call_java("toString")
```
so I can have 
```
    >>> gm = GaussianMixture(k=3, tol=0.0001, seed=10)
    >>> model = gm.fit(df)
    >>> model.setPredictionCol("newPrediction")
    GaussianMixtureModel...
```

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 1]
```diff
@@ -215,6 +215,13 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
     defaultCopy[Bucketizer](extra).setParent(parent)
   }
+
+  @Since("3.0.0")
+  override def toString: String = {
+    s"Bucketizer: uid=$uid" +
+      get(inputCols).map(c => s", numInputCols=${c.length}").getOrElse("") +
+      get(outputCols).map(c => s", numOutputCols=${c.length}").getOrElse("")
+  }
 }
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I test this in REPL:
```
scala> val binarizer: Binarizer = new Binarizer()
binarizer: org.apache.spark.ml.feature.Binarizer = Binarizer: uid=binarizer_3eab2c2d88f6

scala> binarizer.setInputCols(Array("a", "b"))
res3: binarizer.type = Binarizer: uid=binarizer_3eab2c2d88f6, numInputCols=2

scala> binarizer.setOutputCols(Array("c", "d"))
res4: binarizer.type = Binarizer: uid=binarizer_3eab2c2d88f6, numInputCols=2, numOutputCols=2
```

**Comment by zhengruifeng** [comment]
@srowen Yes, I updated them in the second commit.
BTW, I added toString method for evluatores/tunining/features impls

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
@huaxingao Yes, the py side should be updated too.

**Review by srowen — APPROVED** [review]
Looks OK pending tests.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by zhengruifeng** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala [hunk 2]
```diff
@@ -91,6 +91,9 @@ class BisectingKMeansModel private[ml] (
   extends Model[BisectingKMeansModel] with BisectingKMeansParams with MLWritable
   with HasTrainingSummary[BisectingKMeansSummary] {
 
+  @Since("3.0.0")
+  val numFeatures: Int = parentModel.clusterCenters.head.size
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: This seems to be beyond the scope of this PR, but it looks okay. Maybe, `lazy val` is better?

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 3]
```diff
@@ -89,6 +89,9 @@ class GaussianMixtureModel private[ml] (
   extends Model[GaussianMixtureModel] with GaussianMixtureParams with MLWritable
   with HasTrainingSummary[GaussianMixtureSummary] {
 
+  @Since("3.0.0")
+  val numFeatures: Int = gaussians.head.mean.size
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: This PR seems to add at least 4 `numFeatures` instance. Could you add this into the PR description explicitly?

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala [hunk 4]
```diff
@@ -159,6 +159,12 @@ final class VectorSlicer @Since("1.5.0") (@Since("1.5.0") override val uid: Stri
 
   @Since("1.5.0")
   override def copy(extra: ParamMap): VectorSlicer = defaultCopy(extra)
+
+  @Since("3.0.0")
+  override def toString: String = {
+    s"VectorSlicer: uid=$uid," +
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala]: Please remove the ending `,`. We should not have this here.

**Comment by zhengruifeng** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you, @zhengruifeng and @srowen .

**Comment by zhengruifeng** [comment]
Thanks for reviewing! @dongjoon-hyun  @srowen

**Pull request closed without merging**